# Week 7

## Pandas

    import pandas
        pandas.function
    import pandas as pd
        pandas.function

### Series

    a list


In [394]:
import pandas as pd     # import pandas library

dataSeries = ['F350', 'Jeep', 'Ram', 'Motorcycle'] # create a list

mySeries = pd.Series(dataSeries)    # create a Series from the list

print(mySeries)  # print the Series

0          F350
1          Jeep
2           Ram
3    Motorcycle
dtype: object


### Dataframe

    what we use to access data. matrix.

In [395]:
myDataFrame = [('F350', 2), ('Jeep', 10), ('Ram', 25)]  # create a list of tuples

mydf = pd.DataFrame(myDataFrame)

mydf

0   1
0  F350   2
1  Jeep  10
2   Ram  25

### Reading from CSV

In [396]:
# import pandas as pd     # import pandas library

# df = pd.read_csv('../DataFiles/titanic.csv')   # read the CSV file into a DataFrame from local file

In [397]:
import pandas as pd     # import pandas library

df = pd.read_csv('http://128.198.162.191/dataFiles/TITANIC/titanic.csv')   # read the CSV file into a DataFrame from URL

In [398]:
df.info()   # get a summary of the DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [399]:
df.describe()   # get statistical summary of numerical columns

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [400]:
df.head()  # display the first 5 rows of the DataFrame

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [401]:
df.head(6)  # display the first 6 rows of the DataFrame

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q

In [402]:
df[['Name', 'Age']]  # display only the 'Name' and 'Age' columns

Name   Age
0                              Braund, Mr. Owen Harris  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0
2                               Heikkinen, Miss. Laina  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
4                             Allen, Mr. William Henry  35.0
..                                                 ...   ...
886                              Montvila, Rev. Juozas  27.0
887                       Graham, Miss. Margaret Edith  19.0
888           Johnston, Miss. Catherine Helen "Carrie"   NaN
889                              Behr, Mr. Karl Howell  26.0
890                                Dooley, Mr. Patrick  32.0

[891 rows x 2 columns]

### .loc function

    df.loc[args]

In [403]:
df.loc[df['Age'] < 20]  # display rows where 'Age' is less than 20

PassengerId  Survived  Pclass                                  Name  \
7              8         0       3        Palsson, Master. Gosta Leonard   
9             10         1       2   Nasser, Mrs. Nicholas (Adele Achem)   
10            11         1       3       Sandstrom, Miss. Marguerite Rut   
14            15         0       3  Vestrom, Miss. Hulda Amanda Adolfina   
16            17         0       3                  Rice, Master. Eugene   
..           ...       ...     ...                                   ...   
855          856         1       3            Aks, Mrs. Sam (Leah Rosen)   
869          870         1       3       Johnson, Master. Harold Theodor   
875          876         1       3      Najib, Miss. Adele Kiamie "Jane"   
877          878         0       3                  Petroff, Mr. Nedelio   
887          888         1       1          Graham, Miss. Margaret Edith   

        Sex   Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
7      male   2.0      3      1   349909  21.0750   NaN        S  
9    female  14.0      1      0   237736  30.0708   NaN        C  
10   female   4.0      1      1  PP 9549  16.7000    G6        S  
14   female  14.0      0      0   350406   7.8542   NaN        S  
16     male   2.0      4      1   382652  29.1250   NaN        Q  
..      ...   ...    ...    ...      ...      ...   ...      ...  
855  female  18.0      0      1   392091   9.3500   NaN        S  
869    male   4.0      1      1   347742  11.1333   NaN        S  
875  female  15.0      0      0     2667   7.2250   NaN        C  
877    male  19.0      0      0   349212   7.8958   NaN        S  
887  female  19.0      0      0   112053  30.0000   B42        S  

[164 rows x 12 columns]

In [404]:
df.loc[df['Name'] == 'Rice, Master. Eugene']  # display rows where 'Name' matches exactly

PassengerId  Survived  Pclass                  Name   Sex  Age  SibSp  \
16           17         0       3  Rice, Master. Eugene  male  2.0      4   

    Parch  Ticket    Fare Cabin Embarked  
16      1  382652  29.125   NaN        Q

In [405]:
df.loc[df['Name'].str.contains('Brown', case=False)]  # display rows where 'Name' contains 'Brown', case insensitive

PassengerId  Survived  Pclass  \
194          195         1       1   
328          329         1       3   
345          346         1       2   
670          671         1       2   
684          685         0       2   

                                                  Name     Sex   Age  SibSp  \
194          Brown, Mrs. James Joseph (Margaret Tobin)  female  44.0      0   
328     Goldsmith, Mrs. Frank John (Emily Alice Brown)  female  31.0      1   
345                      Brown, Miss. Amelia "Mildred"  female  24.0      0   
670  Brown, Mrs. Thomas William Solomon (Elizabeth ...  female  40.0      1   
684                  Brown, Mr. Thomas William Solomon    male  60.0      1   

     Parch    Ticket     Fare Cabin Embarked  
194      0  PC 17610  27.7208    B4        C  
328      1    363291  20.5250   NaN        S  
345      0    248733  13.0000   F33        S  
670      1     29750  39.0000   NaN        S  
684      1     29750  39.0000   NaN        S

### .iloc

    uses index location

In [406]:
df.head(1)

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1         0       3  Braund, Mr. Owen Harris  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked  
0      0  A/5 21171  7.25   NaN        S

In [407]:
df.iloc[0:3, 0:5]  # display the first 3 rows and first 5 columns using iloc

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex  
0                            Braund, Mr. Owen Harris    male  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  
2                             Heikkinen, Miss. Laina  female

In [408]:
df.iloc[:, 0:6]  # display all rows and first 6 columns using iloc

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  
0                              Braund, Mr. Owen Harris    male  22.0  
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  
2                               Heikkinen, Miss. Laina  female  26.0  
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  
4                             Allen, Mr. William Henry    male  35.0  
..                                                 ...     ...   ...  
886                              Montvila, Rev. Juozas    male  27.0  
887                       Graham, Miss. Margaret Edith  female  19.0  
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN  
889                              Behr, Mr. Karl Howell    male  26.0  
890                                Dooley, Mr. Patrick    male  32.0  

[891 rows x 6 columns]

### Slicing data

In [409]:
df_sub1 = df[['Name', 'Sex', 'Age', 'Survived']]    # create a subset DataFrame with selected columns

df_sub1.head()   # display the first 5 rows of the subset DataFrame

Name     Sex   Age  Survived
0                            Braund, Mr. Owen Harris    male  22.0         0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0         1
2                             Heikkinen, Miss. Laina  female  26.0         1
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0         1
4                           Allen, Mr. William Henry    male  35.0         0

## Titanic Examples

In [410]:
import pandas as pd     # import pandas library
df = pd.read_csv('http://128.198.162.191/dataFiles/TITANIC/titanic.csv')   # read the CSV file into a DataFrame from URL

In [411]:
df_sub1 = df[['Name', 'Sex', 'Age', 'Survived']]    # create a subset DataFrame with selected columns

df_sub1.head()   # display the first 5 rows of the subset DataFrame

Name     Sex   Age  Survived
0                            Braund, Mr. Owen Harris    male  22.0         0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0         1
2                             Heikkinen, Miss. Laina  female  26.0         1
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0         1
4                           Allen, Mr. William Henry    male  35.0         0

In [412]:
df_sub1.query('Survived == True') # filter rows where 'Survived' is True

Name     Sex   Age  Survived
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0         1
2                               Heikkinen, Miss. Laina  female  26.0         1
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0         1
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0         1
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0         1
..                                                 ...     ...   ...       ...
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0         1
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0         1
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0         1
887                       Graham, Miss. Margaret Edith  female  19.0         1
889                              Behr, Mr. Karl Howell    male  26.0         1

[342 rows x 4 columns]

### Handle missing data

In [413]:
df['Survived'].isna()  # check for missing values in 'Survived' column False = not missing, True = missing
df['Age'].isna()  # check for missing values in 'Age' column False = not missing, True = missing

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [414]:
df['Age'].isna().sum()  # count of missing values in 'Age' column

np.int64(177)

In [415]:
df.isna().sum()  # count of missing values in each column of the DataFrame

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### Remove records that have nulls

In [416]:
df_noNull = df.loc[~df['Age'].isna()]  # create a new DataFrame excluding rows with nulls in 'Age' column

df_noNull['Age'].isna() # verify there are no missing values in 'Age' column of the new DataFrame

0      False
1      False
2      False
3      False
4      False
       ...  
885    False
886    False
887    False
889    False
890    False
Name: Age, Length: 714, dtype: bool

In [417]:
df_noNull.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [418]:
df_noNull['Age'].astype('int')  # convert 'Age' column to integer type

0      22
1      38
2      26
3      35
4      35
       ..
885    39
886    27
887    19
889    26
890    32
Name: Age, Length: 714, dtype: int64

In [419]:
df_noNull = df.loc[~df['Age'].isna().copy()]  # create a new DataFrame excluding rows with nulls in 'Age' column

df_noNull['Age'] = df_noNull['Age'].astype('int')  # convert 'Age' column to integer type

df_noNull.dtypes

C:\Users\marja\AppData\Local\Temp\ipykernel_19512\3460028104.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_noNull['Age'] = df_noNull['Age'].astype('int')  # convert 'Age' column to integer type


PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age              int64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

### Ratio of 'Fare' and 'Survived'

In [420]:
df.head(1)

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1         0       3  Braund, Mr. Owen Harris  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked  
0      0  A/5 21171  7.25   NaN        S

In [421]:
df['SurvivalOdds'] = df['Survived'] / df['Fare']  # create a new column 'SurvivalOdds' as the ratio of 'Survived' to 'Fare

df.head()  # display the first 5 rows of the DataFrame with the new column

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  SurvivalOdds  
0      0         A/5 21171   7.2500   NaN        S      0.000000  
1      0          PC 17599  71.2833   C85        C      0.014029  
2      0  STON/O2. 3101282   7.9250   NaN        S      0.126183  
3      0            113803  53.1000  C123        S      0.018832  
4      0            373450   8.0500   NaN        S      0.000000

In [422]:
# df.to_csv('../DataFiles/odds.csv')  # save the modified DataFrame to a new CSV file

In [423]:
survival_counts = df['Survived'].value_counts()  # count occurrences of each unique value in 'Survived' column

print(survival_counts)  # print the counts of survivors and non-survivors

Survived
0    549
1    342
Name: count, dtype: int64


### Overall ratio of survival

In [424]:
survival_prop = df['Survived'].mean()  # calculate the proportion of survivors

print(f'Proportion of survivors: {survival_prop:.2f}')  # print the proportion of survivors formatted to 2 decimal places

Proportion of survivors: 0.38


In [425]:
df['Survived'] = df['Survived'].map({False: 'No', True: 'Yes'})  # map boolean values to strings in 'Survived' column
df.head()

PassengerId Survived  Pclass  \
0            1      NaN       3   
1            2      NaN       1   
2            3      NaN       3   
3            4      NaN       1   
4            5      NaN       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  SurvivalOdds  
0      0         A/5 21171   7.2500   NaN        S      0.000000  
1      0          PC 17599  71.2833   C85        C      0.014029  
2      0  STON/O2. 3101282   7.9250   NaN        S      0.126183  
3      0            113803  53.1000  C123        S      0.018832  
4      0            373450   8.0500   NaN        S      0.000000

### Number of passenegers by class

In [426]:
class_counts = df['Pclass'].value_counts().sort_index()  # count occurrences of each unique value in 'Pclass' column

class_counts = df['Pclass'].value_counts().sort_index(ascending = False)  # count occurrences of each unique value in 'Pclass' column

print(class_counts)  # print the counts of each passenger class

Pclass
3    491
2    184
1    216
Name: count, dtype: int64


### Average amount paid per class

In [427]:
df.groupby('Pclass')['Fare'].mean()  # calculate the mean fare paid per passenger class

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

In [428]:
df.groupby('Pclass')['Fare'].median()  # calculate the median fare paid per passenger class

Pclass
1    60.2875
2    14.2500
3     8.0500
Name: Fare, dtype: float64

### Average age of passengers

    we are missing 177 records

In [429]:
df.groupby('Survived')['Age'].mean()  # calculate the mean age of survivors and non-survivors

Series([], Name: Age, dtype: float64)

### Odds of surviving by class

In [430]:
class_survival_counts = df.groupby('Pclass')['Survived'].value_counts().unstack()  # count survivors and non-survivors by passenger class

odds_of_survival = class_survival_counts[1] / class_survival_counts[0]  # calculate odds of survival for each passenger class

KeyError: 1

### Percentage of survial by passenegr class

In [ ]:
survival_rate = df.groupby('Pclass')['Survived'].mean() * 100  # calculate survival rate by passenger class
print(survival_rate)  # print the survival rate by passenger class

### Plot survival rate by class

In [ ]:
import matplotlib.pyplot as plt  # import matplotlib for plotting

survival_rate.plot(kind = 'bar', rot = 0)

plt.xlabel('Passenger Class')  # label for x-axis
plt.ylabel('Survival Rate (%)')  # label for y-axis
plt.title('Survival Rate by Passenger Class')  # title of the plot
plt.show()  # display the plot